### DISCUSSION TITLE
cartopy version

### DISCUSSION THREAD

**Discussion by avaldebe** [metno/pyaerocom#583] [General]
Currently we can not use cartopy 0.20 on CI as it [requires Proj 8+](https://github.com/SciTools/iris/issues/4468) and Ubuntu 20.04LTS only provides [6.3.1](https://packages.ubuntu.com/focal/proj-bin).

In order to `pip install "cartopy>=0.20"`, one would need to
- update to the next LTS (Ubuntu 22.04), which will pack [Proj 8.2.1](https://packages.ubuntu.com/jammy/proj-bin)
- build Proj from source

None of these options are reasonable for CI, not to users installing from pip (conda users are not affected).

Also, [Iris 3.2 requires cartopy >=0.20](https://github.com/SciTools/iris/blob/main/setup.cfg#L48)

**Comment by avaldebe** [comment]
as far as I can tell we use `cartopy` on
- `pyaerocom.region:Region.plot`
- `pyaerocom.aeroval.modelmaps_helpers:calc_contour_json`
- `pyaerocom.plot.helpers:projection_from_str`
- `pyaerocom.plot.mapping:set_map_ticks`
- `pyaerocom.plot.mapping:init_map`
- `pyaerocom.plot.mapping:plot_griddeddata_on_map`
- `pyaerocom.plot.mapping:plot_nmb_map_colocateddata`
- `pyaerocom.plot.plotcoordinates:plot_coordinates`

**Reply by jgriesfeller** [reply]
Shouldn't it be possible to build proj8 and put the resulting libraries and support files into a tar file that CI then installs during setup of the CI VMs? 
Untaring e.g. to ${HOME}/lib and adding the path to $LD_LIBRARY_PATH should be enough for that.

**Reply by avaldebe** [reply]
it would be "easier" to use a custom docker container, but we would not be testing against the environment of a pip user

**Comment by avaldebe** [comment]
Cartopy version 0.19.* does no longer install on CI, e.g. [here][docs job] and [here][lint job]

[docs job]: https://github.com/metno/pyaerocom/runs/5689076088?check_suite_focus=true
[lint job]: https://github.com/metno/pyaerocom/runs/5689075713?check_suite_focus=true

The problem is that `cartopy` [setup script] imports `convert_path` from `setuptools`

[setup script]: https://github.com/SciTools/cartopy/blob/v0.19.x/setup.py#L35

``` Python
from setuptools import Command, Extension, convert_path, setup
```

but `convert_path` is not provided by `setuptools`.

Up to version [60.9.3], `setuptools` imports `convert_path` from `distutils.util`.

``` Python
from distutils.util import convert_path
```

[60.9.3]: https://github.com/pypa/setuptools/blob/v60.9.3/setuptools/__init__.py#L12
[60.10.0]: https://github.com/pypa/setuptools/blob/v60.10.0/setuptools/__init__.py

Now, as `distutils` was deprecated on Python 3.10 and will be removed on Python 3.12 (see [PEP 632]),
`setuptools` cleaned up the dependency on 60.10.0 which was released on [March 16th][60.10.0]...

[PEP 632]: https://peps.python.org/pep-0632/#specification

**Comment by jgriesfeller** [comment]
I had a look at the cartopy github pages. In the current master this problem has been solved (commit from Friday):
https://github.com/SciTools/cartopy/blob/c9be040cdc015f15289433a3cd7bcd049bb57a9f/setup.py#L34

Not sure when they will release the next time though.

**Reply by avaldebe** [reply]
This would affect the next version `0.20.*`, which needs Proj 8+.
On CI we need version `0.19.*`, which uses needs Proj 4.9+.

In any case, I can open an issue asking for a back-port of the fix and a new `0.19.0.post2` release.

**Comment by magnusuMET** [comment]
`0.19` series is not forwards compatible either, and the pinning in `setup.cfg` means I can not install on rhel8. Is there a way to pin `cartopy==0.19` in CI and removing it from setup.cfg?

**Reply by avaldebe** [reply]
tried different things, but the only reliable way was to move cartopy/iris to extras, PR #620
